New integration: Hue BLE

flip-dots opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

This integration allows for the newer Phillips Hue lights with Bluetooth to 
be controlled over Bluetooth by Home Assistant instead of having to use 
a Zigbee dongle. I think support for this is quite important because many 
people (myself at least) trialed Home Assistant on hardware they have 
lying around like a raspberry pi, old laptop, or mini pc, which have 
Bluetooth built in but no Zigbee support and don't want to spend money 
on hardware they don't necessarily need, at least without knowing that its 
worth it.

The features are as follows:
- On/Off control
- Brightness control
- Color temp control
- XY color control
- Light state reading
  - Power
  - Brightness
  - Color temp
  - XY color
- Light metadata reading
  - Name
  - Manufacturer
  - Model
  - Firmware version
  - Zigbee address
- Push approach


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

This is my first real integration for Home Assistant and also the first
Python library I have programmed so I have probably made a few mistakes
but I've been testing it for the past few months and have ironed out the issues
as they have appeared, I've been using the latest version for the past two 
weeks and haven't gotten any errors or reliability issues, yet anyway...

The library can be found [here](https://github.com/flip-dots/HueBLE) and the docs [here](https://hueble.readthedocs.io/en/latest/index.html)

I only have the LCA006 model which is 1100 lumens which does RGB and
WW/CW so that is the only model confirmed to work but I have tried to design
the library and integration to detect the supported features and set the flags
accordingly but this is untested on other models, if anyone happens to have
one of the other models that supports Bluetooth and could test it for me that
would be great.

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/33042

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] I have followed the [perfect PR recommendations][perfect-pr]
- [X] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [X] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [X] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [X] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [X] Untested files have been added to `.coveragerc`.

[library changelog](https://hueble.readthedocs.io/en/latest/history.html)

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

flip-dots committed:
Initial implementation of hue_ble.

flip-dots committed:
Update deps

flip-dots committed:
Update deps

flip-dots committed:
Resolve issue with unsupported light mode warning.

flip-dots committed:
Update deps

flip-dots committed:
Add tests for hue_ble

flip-dots committed:
Update codeowners

flip-dots committed:
Update strings to be more specific.

flip-dots committed:
Update deps

zweckj commented on homeassistant/components/hue_ble/config_flow.py:
```diff
@@ -0,0 +1,164 @@
+"""Config flow for Hue BLE integration."""
```
config flow needs to have 100% coverage

flip-dots committed:
Add error messages to bluetooth discovery flow

flip-dots committed:
Increase test coverage of config flow

tr4nt0r commented on homeassistant/components/hue_ble/config_flow.py:
```diff
@@ -0,0 +1,186 @@
+            _LOGGER.exception("Unexpected exception")
+            errors["base"] = "unknown"
+        else:
+            return self.async_create_entry(title=user_input[CONF_NAME], data=user_input)
+
```
```suggestion
            return self.async_create_entry(title=user_input[CONF_NAME], data={})

```

We don't need to store name and mac in data, the name is stored in entry title, mac address is the unique_id

tr4nt0r commented on homeassistant/components/hue_ble/light.py:
```diff
@@ -0,0 +1,185 @@
+
+        _LOGGER.debug("Received state notification from light %s", self.name)
+
+        self._name = self._light.name
+        self._address = self._light.address
+        self._attr_available = self._light.available
+        self._attr_name = self._name
+        self._attr_is_on = self._light.power_state
+        self._attr_brightness = self._light.brightness
+        self._attr_color_temp = self._light.colour_temp
+        self._attr_xy_color = self._light.colour_xy
+
```
You should use properties for all these value updates instead:

```
@property
def is_on(self) -> bool:
    """Return true if device is on."""
    return self._light.power_state
```

Also, don't update the name

tr4nt0r commented:
The integration also fails with an uncatched exception. That might be an error in the library. If you can't fix this in the library you might want to catch this exception in the config flow

```
2024-06-29 19:52:37.239 ERROR (MainThread) [HueBLE] Exception connecting to light "Hue filament bulb". Error "'NoneType' object has no attribute 'get'".
2024-06-29 19:52:37.442 INFO (MainThread) [HueBLE] Received expected disconnect from "HaBleakClientWrapper, xx:xx:xx:xx:xx:xx".
2024-06-29 19:52:37.443 ERROR (MainThread) [homeassistant.components.hue_ble.config_flow] Unexpected exception
Traceback (most recent call last):
  File "/workspaces/ha-core/homeassistant/components/hue_ble/config_flow.py", line 147, in async_step_confirm
    await validate_input(self.hass, user_input)
  File "/workspaces/ha-core/homeassistant/components/hue_ble/config_flow.py", line 50, in validate_input
    if not light.authenticated:
           ^^^^^^^^^^^^^^^^^^^
  File "/home/vscode/.local/ha-venv/lib/python3.12/site-packages/HueBLE.py", line 933, in authenticated
    if properties.get("Paired") is True:
       ^^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'get'
```

flip-dots commented on homeassistant/components/hue_ble/config_flow.py:
```diff
@@ -0,0 +1,186 @@
+        """Handle manually setting up a light."""
+
+        errors: dict[str, str] = {}
+        if user_input is None:
+            return self.async_show_form(
+                step_id="user", data_schema=STEP_USER_DATA_SCHEMA, errors=errors
+            )
+
+        try:
```
I implemented this in an earlier revision but I wasn't able to get it to work reliably, it would produce the list fine but when shown in the UI the entries in the list would be present and select-able but would be blank, i.e no text was shown where the name and mac should have been and I was not able to find a fix and was running out of time I could justify spending on it, especially since it is not the intended way to use the integration, more as a backup in case automatic Bluetooth discovery was not working for some reason.

tr4nt0r replied:
If you don't cache the discovered devices in `self`, they won't carry over to the next step. This example user flow should work with some small adjustments:

```
    async def async_step_user(
        self, user_input: dict[str, Any] | None = None
    ) -> ConfigFlowResult:
        """Handle the user step to pick discovered device."""
        if user_input is not None:
            address = user_input[CONF_ADDRESS]
            title = self._discovered_devices[address]
            await self.async_set_unique_id(address, raise_on_progress=False)
            self._abort_if_unique_id_configured()
            return self.async_create_entry(title=title, data={})

        current_addresses = self._async_current_ids()
        for discovery_info in async_discovered_service_info(self.hass, True):
            address = discovery_info.address
            if (
                DISCOVERY_SVC_UUID not in discovery_info.service_uuids
                or address in current_addresses
                or address in self._discovered_devices
            ):
                continue
            self._discovered_devices[address] = discovery_info.name

        if not self._discovered_devices:
            return self.async_abort(reason="no_devices_found")

        return self.async_show_form(
            step_id="user",
            data_schema=self.add_suggested_values_to_schema(vol.Schema(
                {vol.Required(CONF_ADDRESS): vol.In(self._discovered_devices)}
            ), user_input
            ),
        )
```

flip-dots replied:
I implemented this in a test version but it highlighted a problem I encountered previously but forgotten, using this manual method over the Bluetooth discovery method does not work if the Bluetooth discovery step detects the light since that flow will already be in progress, meaning the manual flow with the name and MAC fields will not be able to progress, which makes automatic filling of the form kind of pointless, I guess I could modify it so Bluetooth discovery does not set the ID and use some other mechanism to avoid duplicates but that seems overkill, maybe just edit the strings to point out the manual form with the name and MAC should only be used if the automatic Bluetooth discovery doesn't detect it?

Here is the code I added to test this:
``` 
       errors: dict[str, str] = {}
        if user_input is None:
            current_addresses = self._async_current_ids()
            for discovery_info in async_discovered_service_info(self.hass, True):
                address = discovery_info.address
                if (
                    UUID_HUE_IDENTIFIER in discovery_info.service_uuids
                    and dr.format_mac(address) not in current_addresses
                ):  # or address in self._discovered_devices
                    return self.async_show_form(
                        step_id="user",
                        data_schema=self.add_suggested_values_to_schema(
                            STEP_USER_DATA_SCHEMA,
                            {CONF_NAME: discovery_info.name, CONF_MAC: address},
                        ),
                        description_placeholders={"url_pairing_mode": URL_PAIRING_MODE},
                        errors=errors,
                    )
```

flip-dots committed:
Replace usage of redundant form data in config entry

flip-dots committed:
Replace hass.data with entry.runtime_data

flip-dots committed:
Remove redundant device info

flip-dots committed:
Remove redundant attr

flip-dots committed:
Add light entity description

flip-dots committed:
Remove (mostly) static state variables from state updates

flip-dots committed:
Replace usage of attr state update with properties

flip-dots committed:
Update intergration name

flip-dots committed:
Update deps

flip-dots committed:
Add support for light authentication status being unknown

flip-dots commented on homeassistant/components/hue_ble/__init__.py:
```diff
@@ -0,0 +1,59 @@
+    ble_device = async_ble_device_from_address(hass, address.upper(), connectable=True)
+
+    if ble_device is None:
+        count_scanners = async_scanner_count(hass, connectable=True)
+        _LOGGER.debug("Count of BLE scanners: %i", count_scanners)
+
+        if count_scanners < 1:
+            raise ConfigEntryNotReady(
+                "No Bluetooth scanners are available to search for the light."
+            )
+        raise ConfigEntryNotReady("The light was not found.")
```
I saw this in one of the integrations I based this on and I kept it because I'd rather the error for not being able to find it and not being able to find it because there is nothing available to search for it be different things.

emontnemery replied:
I think this is somewhat valid, but I also think this kind of check should live in some BLE integration helper so not every integration has to duplicate such code.
@bdraco, do we have some helper like this already?

flip-dots commented:
> The integration also fails with an uncatched exception. That might be an error in the library. If you can't fix this in the library you might want to catch this exception in the config flow
> 
> ```
> 2024-06-29 19:52:37.239 ERROR (MainThread) [HueBLE] Exception connecting to light "Hue filament bulb". Error "'NoneType' object has no attribute 'get'".
> 2024-06-29 19:52:37.442 INFO (MainThread) [HueBLE] Received expected disconnect from "HaBleakClientWrapper, xx:xx:xx:xx:xx:xx".
> 2024-06-29 19:52:37.443 ERROR (MainThread) [homeassistant.components.hue_ble.config_flow] Unexpected exception
> Traceback (most recent call last):
>   File "/workspaces/ha-core/homeassistant/components/hue_ble/config_flow.py", line 147, in async_step_confirm
>     await validate_input(self.hass, user_input)
>   File "/workspaces/ha-core/homeassistant/components/hue_ble/config_flow.py", line 50, in validate_input
>     if not light.authenticated:
>            ^^^^^^^^^^^^^^^^^^^
>   File "/home/vscode/.local/ha-venv/lib/python3.12/site-packages/HueBLE.py", line 933, in authenticated
>     if properties.get("Paired") is True:
>        ^^^^^^^^^^^^^^
> AttributeError: 'NoneType' object has no attribute 'get'
> ```

Seems like my method for detecting if a device is authenticated doesn't work for all Linux systems, (It works fine on HAOS and Mint tho which is strange), I updated the library so the method returns None if it cannot determine the authentication status on a Linux system and updated the integration to ignore if it can't tell if its authenticated and proceed anyway.

tr4nt0r commented on tests/components/hue_ble/test_config_flow.py:
```diff
@@ -0,0 +1,627 @@
+        )
+        await hass.async_block_till_done()
+
+    assert result["type"] == FlowResultType.CREATE_ENTRY
```
```suggestion
    assert result["type"] is FlowResultType.CREATE_ENTRY
```

tr4nt0r commented on tests/components/hue_ble/test_config_flow.py:
```diff
@@ -0,0 +1,627 @@
+async def test_form(hass: HomeAssistant, mock_setup_entry: AsyncMock) -> None:
+    """Test form with good data."""
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
```
```suggestion
        DOMAIN, context={"source": SOURCE_USER}
```

And import like this instead:

`from homeassistant.config_entries import SOURCE_USER`

tr4nt0r commented on tests/components/hue_ble/test_config_flow.py:
```diff
@@ -0,0 +1,627 @@
+    assert result["type"] == FlowResultType.FORM
+    assert result["errors"] == {}
+
+    with (
+        patch(
+            "homeassistant.components.hue_ble.config_flow.async_ble_device_from_address",
+            return_value=generate_ble_device(TEST_DEVICE_NAME, TEST_DEVICE_MAC),
+        ),
+        patch(
+            "homeassistant.components.hue_ble.config_flow.async_scanner_count",
+            return_value=1,
+        ),
+        patch(
+            "homeassistant.components.hue_ble.config_flow.HueBleLight.connect",
+            return_value=True,
+        ),
+        patch(
+            "homeassistant.components.hue_ble.config_flow.HueBleLight.authenticated",
+            return_value=True,
+            new_callable=PropertyMock,
+        ),
+        patch(
+            "homeassistant.components.hue_ble.config_flow.HueBleLight.connected",
+            return_value=True,
+            new_callable=PropertyMock,
+        ),
+        patch(
+            "homeassistant.components.hue_ble.config_flow.HueBleLight.poll_state",
+            return_value=(True, []),
+        ),
+    ):
+        result = await hass.config_entries.flow.async_configure(
```
Move this to a fixture in conftest.py, and use it for all tests to avoid repetition

tr4nt0r commented on tests/components/hue_ble/test_config_flow.py:
```diff
@@ -0,0 +1,627 @@
+        )
+        await hass.async_block_till_done()
+
+    assert result["type"] == FlowResultType.CREATE_ENTRY
```
```suggestion
    assert result["type"] is FlowResultType.CREATE_ENTRY
```

tr4nt0r commented on tests/components/hue_ble/test_config_flow.py:
```diff
@@ -0,0 +1,627 @@
+            {CONF_NAME: TEST_DEVICE_NAME, CONF_MAC: TEST_DEVICE_MAC},
+        )
+
+    assert result["type"] == FlowResultType.FORM
```
```suggestion
    assert result["type"] is FlowResultType.FORM
```

tr4nt0r commented on tests/components/hue_ble/test_config_flow.py:
```diff
@@ -0,0 +1,627 @@
+        )
+        await hass.async_block_till_done()
+
+    assert result["type"] == FlowResultType.CREATE_ENTRY
```
```suggestion
    assert result["type"] is FlowResultType.CREATE_ENTRY
```

tr4nt0r commented on tests/components/hue_ble/test_config_flow.py:
```diff
@@ -0,0 +1,627 @@
+            {CONF_NAME: TEST_DEVICE_NAME, CONF_MAC: TEST_DEVICE_MAC},
+        )
+
+    assert result["type"] == FlowResultType.FORM
```
```suggestion
    assert result["type"] is FlowResultType.FORM
```

tr4nt0r commented on tests/components/hue_ble/test_config_flow.py:
```diff
@@ -0,0 +1,627 @@
+) -> None:
+    """Test form when not authenticated to device (pair failed)."""
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
```
```suggestion
        DOMAIN, context={"source": SOURCE_USER}
```

tr4nt0r commented on tests/components/hue_ble/test_config_flow.py:
```diff
@@ -0,0 +1,627 @@
+        )
+        await hass.async_block_till_done()
+
+    assert result["type"] == FlowResultType.CREATE_ENTRY
```
```suggestion
    assert result["type"] is FlowResultType.CREATE_ENTRY
```

tr4nt0r commented on tests/components/hue_ble/test_config_flow.py:
```diff
@@ -0,0 +1,627 @@
+) -> None:
+    """Test form when authentication status is unknown."""
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
```
```suggestion
        DOMAIN, context={"source": SOURCE_USER}
```

tr4nt0r commented on tests/components/hue_ble/test_config_flow.py:
```diff
@@ -0,0 +1,627 @@
+            {CONF_NAME: TEST_DEVICE_NAME, CONF_MAC: TEST_DEVICE_MAC},
+        )
+
+    assert result["type"] == FlowResultType.CREATE_ENTRY
```
```suggestion
    assert result["type"] is FlowResultType.CREATE_ENTRY
```

tr4nt0r commented on tests/components/hue_ble/test_config_flow.py:
```diff
@@ -0,0 +1,627 @@
+) -> None:
+    """Test form when device connection fails."""
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
```
```suggestion
        DOMAIN, context={"source": SOURCE_USER}
```

tr4nt0r commented on tests/components/hue_ble/test_config_flow.py:
```diff
@@ -0,0 +1,627 @@
+
+    with patch(
+        "homeassistant.components.hue_ble.config_flow.validate_input",
+        side_effect=Exception,
```
Use parametrization instead. This way you will only need 1 test instead of 5 to cover all exceptions

flip-dots committed:
Simplify light color_mode detection

flip-dots committed:
Refactor tests

tr4nt0r commented:
> > The integration also fails with an uncatched exception. That might be an error in the library. If you can't fix this in the library you might want to catch this exception in the config flow
> > ```
> > 2024-06-29 19:52:37.239 ERROR (MainThread) [HueBLE] Exception connecting to light "Hue filament bulb". Error "'NoneType' object has no attribute 'get'".
> > 2024-06-29 19:52:37.442 INFO (MainThread) [HueBLE] Received expected disconnect from "HaBleakClientWrapper, xx:xx:xx:xx:xx:xx".
> > 2024-06-29 19:52:37.443 ERROR (MainThread) [homeassistant.components.hue_ble.config_flow] Unexpected exception
> > Traceback (most recent call last):
> >   File "/workspaces/ha-core/homeassistant/components/hue_ble/config_flow.py", line 147, in async_step_confirm
> >     await validate_input(self.hass, user_input)
> >   File "/workspaces/ha-core/homeassistant/components/hue_ble/config_flow.py", line 50, in validate_input
> >     if not light.authenticated:
> >            ^^^^^^^^^^^^^^^^^^^
> >   File "/home/vscode/.local/ha-venv/lib/python3.12/site-packages/HueBLE.py", line 933, in authenticated
> >     if properties.get("Paired") is True:
> >        ^^^^^^^^^^^^^^
> > AttributeError: 'NoneType' object has no attribute 'get'
> > ```
> 
> Seems like my method for detecting if a device is authenticated doesn't work for all Linux systems, (It works fine on HAOS and Mint tho which is strange), I updated the library so the method returns None if it cannot determine the authentication status on a Linux system and updated the integration to ignore if it can't tell if its authenticated and proceed anyway.

I'm still not able to connect to my hue lamp so i can't really test. (I'm getting error code 82), would be good if you work on making the library more robust and test it more thoroughly with different setups. I am testing it in the Home Assistant  Devcontainer (on Windows with WSL2) and an ESPHome Bluetooth proxy.
)

tr4nt0r commented on tests/components/hue_ble/test_config_flow.py:
```diff
@@ -0,0 +1,291 @@
+    assert result["step_id"] == "confirm"
+
+    with patch(
+        "homeassistant.components.hue_ble.config_flow.validate_input",
```
same

tr4nt0r commented on tests/components/hue_ble/test_config_flow.py:
```diff
@@ -0,0 +1,291 @@
+        )
+        await hass.async_block_till_done()
+
+    assert result["type"] == FlowResultType.FORM
```
```suggestion
    assert result["type"] is FlowResultType.FORM
```

tr4nt0r commented on homeassistant/components/hue_ble/strings.json:
```diff
@@ -0,0 +1,26 @@
+  "config": {
+    "step": {
+      "user": {
+        "description": "Enter a name for the device and the MAC address of the light. Make sure the light is in Alexa/Google pairing mode.",
```
You could provide a link for further information on how to enable the pairing mode, either to a Phillips Hue page or a link with anchor to the documentation of the integration (provide the url as placeholder, links should not go into the translation itself)
```suggestion
        "description": "Enter a name for the device and the MAC address of the light. Make sure the light is in [Alexa/Google pairing mode.]({url_pairing_mode})",
```

tr4nt0r commented on homeassistant/components/hue_ble/light.py:
Add files that are not covered by tests to the .coveragerc file, same goes for the __init__.py, or even better, add tests to cover the init and the light platform

frenck commented:
I've marked this PR, as changes are requested that need to be processed.
Please un-draft it once it is ready for review again by clicking the **"Ready for review"** button.

Thanks! 👍 

../Frenck

[_Learn more about our pull request process._](https://developers.home-assistant.io/docs/review-process#prs-are-being-drafted-when-changes-are-needed)

flip-dots commented:
> > > The integration also fails with an uncatched exception. That might be an error in the library. If you can't fix this in the library you might want to catch this exception in the config flow
> > > ```
> > > 2024-06-29 19:52:37.239 ERROR (MainThread) [HueBLE] Exception connecting to light "Hue filament bulb". Error "'NoneType' object has no attribute 'get'".
> > > 2024-06-29 19:52:37.442 INFO (MainThread) [HueBLE] Received expected disconnect from "HaBleakClientWrapper, xx:xx:xx:xx:xx:xx".
> > > 2024-06-29 19:52:37.443 ERROR (MainThread) [homeassistant.components.hue_ble.config_flow] Unexpected exception
> > > Traceback (most recent call last):
> > >   File "/workspaces/ha-core/homeassistant/components/hue_ble/config_flow.py", line 147, in async_step_confirm
> > >     await validate_input(self.hass, user_input)
> > >   File "/workspaces/ha-core/homeassistant/components/hue_ble/config_flow.py", line 50, in validate_input
> > >     if not light.authenticated:
> > >            ^^^^^^^^^^^^^^^^^^^
> > >   File "/home/vscode/.local/ha-venv/lib/python3.12/site-packages/HueBLE.py", line 933, in authenticated
> > >     if properties.get("Paired") is True:
> > >        ^^^^^^^^^^^^^^
> > > AttributeError: 'NoneType' object has no attribute 'get'
> > > ```
> > 
> > 
> > Seems like my method for detecting if a device is authenticated doesn't work for all Linux systems, (It works fine on HAOS and Mint tho which is strange), I updated the library so the method returns None if it cannot determine the authentication status on a Linux system and updated the integration to ignore if it can't tell if its authenticated and proceed anyway.
> 
> I'm still not able to connect to my hue lamp so i can't really test. (I'm getting error code 82), would be good if you work on making the library more robust and test it more thoroughly with different setups. I am testing it in the Home Assistant Devcontainer (on Windows with WSL2) and an ESPHome Bluetooth proxy. )

That authentication error makes sense now, I never designed this system with the idea of a Bluetooth proxy in mind, I have been developing on Mint using the devcontainer to pass through my laptops Bluetooth card by editing the container file and my Home Assistant server is using a USB dongle. That being said I don't think there should be any issue using a Bluetooth proxy though I don't have one to test, if its important I can buy one to try to get it to work, I think its more likely something to do with that light working slightly differently from the model I have and can use to test, if you have a moment my test library has some [example](https://hueble.readthedocs.io/en/latest/examples.html)  scripts which should work on Windows and could rule that out.

flip-dots commented on tests/components/hue_ble/test_config_flow.py:
```diff
@@ -0,0 +1,291 @@
+from tests.components.bluetooth import generate_ble_device
+
+
+@pytest.mark.parametrize(
+    (
+        "return_device_result",
+        "scanner_count_result",
+        "connect_result",
+        "authenticated_result",
+        "connected_result",
+        "poll_state_result",
+        "expected_error",
+    ),
```
I know its confusing but I don't have any better ideas, the validation method takes a lot of parameters and pytest does not seem to accept defaults for them, and the config flow needs 100% coverage, I could try moving the validation method outside of the config flow and the 100% coverage requirement but that feels like even worse of a solution.

flip-dots commented on tests/components/hue_ble/test_config_flow.py:
```diff
@@ -0,0 +1,291 @@
+        "expected_error",
+    ),
+    [
+        pytest.param(
+            generate_ble_device(TEST_DEVICE_NAME, TEST_DEVICE_MAC),
+            1,
+            True,
+            True,
+            True,
+            (True, []),
+            None,
+            id="good_data",
+        ),
```
I know but doing it this way allows me to use the id parameter to describe what the test is for which also shows when pytest is run.

tr4nt0r replied:
That's not true, `@pytest.mark.parametrize`  has the parameter `ids=`, that takes a list with the ids for each test

flip-dots commented on tests/components/hue_ble/test_config_flow.py:
```diff
@@ -0,0 +1,291 @@
+    assert result["errors"] == {"base": "unknown"}
+
+    with patch(
+        "homeassistant.components.hue_ble.config_flow.validate_input",
```
I could but then I would have to use the same messy parameterization to drive it like for the form test, and since it already has 100% coverage I figured this was cleaner.

tr4nt0r replied:
then it's totally fine for me 👍

flip-dots committed:
Refactor multiple mock patches into one

flip-dots committed:
Replace "==" with "is"

flip-dots committed:
Add URL to docs

Co-authored-by: Mr. Bubbles <manni@zapto.de>

flip-dots committed:
Add link to docs in setup form

tr4nt0r commented on tests/components/hue_ble/test_config_flow.py:
```diff
@@ -0,0 +1,291 @@
+
+@pytest.mark.parametrize(
+    ("side_effect", "error_message"),
+    [
+        pytest.param(CannotConnect, "cannot_connect", id="cannot_connect"),
+        pytest.param(InvalidAuth, "invalid_auth", id="device_not_authenticated"),
+        pytest.param(ScannerNotAvailable, "no_scanners", id="scanner_not_avaliable"),
+        pytest.param(NotFound, "not_found", id="device_not_found"),
+        pytest.param(Exception, "unknown", id="unknown"),
+    ],
```
Updated the suggested change and added ids

flip-dots committed:
Simplify test parametrization

Co-authored-by: Mr. Bubbles <manni@zapto.de>

flip-dots committed:
Simplify test parametrization

flip-dots commented:
> > > > The integration also fails with an uncatched exception. That might be an error in the library. If you can't fix this in the library you might want to catch this exception in the config flow
> > > > ```
> > > > 2024-06-29 19:52:37.239 ERROR (MainThread) [HueBLE] Exception connecting to light "Hue filament bulb". Error "'NoneType' object has no attribute 'get'".
> > > > 2024-06-29 19:52:37.442 INFO (MainThread) [HueBLE] Received expected disconnect from "HaBleakClientWrapper, xx:xx:xx:xx:xx:xx".
> > > > 2024-06-29 19:52:37.443 ERROR (MainThread) [homeassistant.components.hue_ble.config_flow] Unexpected exception
> > > > Traceback (most recent call last):
> > > >   File "/workspaces/ha-core/homeassistant/components/hue_ble/config_flow.py", line 147, in async_step_confirm
> > > >     await validate_input(self.hass, user_input)
> > > >   File "/workspaces/ha-core/homeassistant/components/hue_ble/config_flow.py", line 50, in validate_input
> > > >     if not light.authenticated:
> > > >            ^^^^^^^^^^^^^^^^^^^
> > > >   File "/home/vscode/.local/ha-venv/lib/python3.12/site-packages/HueBLE.py", line 933, in authenticated
> > > >     if properties.get("Paired") is True:
> > > >        ^^^^^^^^^^^^^^
> > > > AttributeError: 'NoneType' object has no attribute 'get'
> > > > ```
> > > 
> > > 
> > > Seems like my method for detecting if a device is authenticated doesn't work for all Linux systems, (It works fine on HAOS and Mint tho which is strange), I updated the library so the method returns None if it cannot determine the authentication status on a Linux system and updated the integration to ignore if it can't tell if its authenticated and proceed anyway.
> > 
> > 
> > I'm still not able to connect to my hue lamp so i can't really test. (I'm getting error code 82), would be good if you work on making the library more robust and test it more thoroughly with different setups. I am testing it in the Home Assistant Devcontainer (on Windows with WSL2) and an ESPHome Bluetooth proxy. )
> 
> That authentication error makes sense now, I never designed this system with the idea of a Bluetooth proxy in mind, I have been developing on Mint using the devcontainer to pass through my laptops Bluetooth card by editing the container file and my Home Assistant server is using a USB dongle. That being said I don't think there should be any issue using a Bluetooth proxy though I don't have one to test, if its important I can buy one to try to get it to work, I think its more likely something to do with that light working slightly differently from the model I have and can use to test, if you have a moment my test library has some [example](https://hueble.readthedocs.io/en/latest/examples.html) scripts which should work on Windows and could rule that out.

I have obtained an ESP32 and set it up as a Bluetooth proxy and it seems to work fine, I think the error must be to do with that specific model of light, its probably controlled using different GATT addresses.

tr4nt0r commented:
> > > > > The integration also fails with an uncatched exception. That might be an error in the library. If you can't fix this in the library you might want to catch this exception in the config flow
> > > > > ```
> > > > > 2024-06-29 19:52:37.239 ERROR (MainThread) [HueBLE] Exception connecting to light "Hue filament bulb". Error "'NoneType' object has no attribute 'get'".
> > > > > 2024-06-29 19:52:37.442 INFO (MainThread) [HueBLE] Received expected disconnect from "HaBleakClientWrapper, xx:xx:xx:xx:xx:xx".
> > > > > 2024-06-29 19:52:37.443 ERROR (MainThread) [homeassistant.components.hue_ble.config_flow] Unexpected exception
> > > > > Traceback (most recent call last):
> > > > >   File "/workspaces/ha-core/homeassistant/components/hue_ble/config_flow.py", line 147, in async_step_confirm
> > > > >     await validate_input(self.hass, user_input)
> > > > >   File "/workspaces/ha-core/homeassistant/components/hue_ble/config_flow.py", line 50, in validate_input
> > > > >     if not light.authenticated:
> > > > >            ^^^^^^^^^^^^^^^^^^^
> > > > >   File "/home/vscode/.local/ha-venv/lib/python3.12/site-packages/HueBLE.py", line 933, in authenticated
> > > > >     if properties.get("Paired") is True:
> > > > >        ^^^^^^^^^^^^^^
> > > > > AttributeError: 'NoneType' object has no attribute 'get'
> > > > > ```
> > > > 
> > > > 
> > > > Seems like my method for detecting if a device is authenticated doesn't work for all Linux systems, (It works fine on HAOS and Mint tho which is strange), I updated the library so the method returns None if it cannot determine the authentication status on a Linux system and updated the integration to ignore if it can't tell if its authenticated and proceed anyway.
> > > 
> > > 
> > > I'm still not able to connect to my hue lamp so i can't really test. (I'm getting error code 82), would be good if you work on making the library more robust and test it more thoroughly with different setups. I am testing it in the Home Assistant Devcontainer (on Windows with WSL2) and an ESPHome Bluetooth proxy. )
> > 
> > 
> > That authentication error makes sense now, I never designed this system with the idea of a Bluetooth proxy in mind, I have been developing on Mint using the devcontainer to pass through my laptops Bluetooth card by editing the container file and my Home Assistant server is using a USB dongle. That being said I don't think there should be any issue using a Bluetooth proxy though I don't have one to test, if its important I can buy one to try to get it to work, I think its more likely something to do with that light working slightly differently from the model I have and can use to test, if you have a moment my test library has some [example](https://hueble.readthedocs.io/en/latest/examples.html) scripts which should work on Windows and could rule that out.
> 
> I have obtained an ESP32 and set it up as a Bluetooth proxy and it seems to work fine, I think the error must be to do with that specific model of light, its probably controlled using different GATT addresses.

Hmm, that's weird, because it is also an LCA006 that I have

tr4nt0r reviewed (approved):
I think that's it. A maintainer surely will request more changes but from my part there is nothing more I can say, LGTM 👍

emontnemery commented on homeassistant/components/hue_ble/__init__.py:
```diff
@@ -0,0 +1,52 @@
+            )
+        raise ConfigEntryNotReady("The light was not found.")
+
+    light = HueBleLight(ble_device)
+
+    if not await light.connect() or not await light.poll_state():
+        raise ConfigEntryNotReady("Device found but unable to connect.")
```
So, we do a lot of checks here to ensure the light can be found when setting up the integration.
What happens if the light is found now, but connection is then lost, does the `HueBLE` library automatically reconnect?

flip-dots replied:
It automatically attempts to reconnect once but if that does not work it marks itself as unavailable, but any command sent to it afterwards will initiate a reconnection attempt.

flip-dots replied:
It looks like I had gotten myself confused with how it worked in an earlier iteration.

The way the library currently works is that if it successfully connects and then becomes disconnected without the disconnect method being called it will attempt to reconnect indefinitely. No command is needed to trigger a re-connection attempt, it does it by itself.

emontnemery commented on homeassistant/components/hue_ble/__init__.py:
```diff
@@ -0,0 +1,52 @@
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
```
Add this:

```py
type HueBLEConfigEntry = ConfigEntry[HueBleLight]

...

async def async_setup_entry(hass: HomeAssistant, entry: HueBLEConfigEntry ) -> bool:

...

async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:

```

emontnemery replied:
You marked this as resolved, but you forgot to update the type annotations of `entry` passed to `async_setup_entry` + `async_unload_entry`, please fix that.

emontnemery commented on homeassistant/components/hue_ble/light.py:
```diff
@@ -0,0 +1,192 @@
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
```
```suggestion
if TYPE_CHECKING:
    from . import HueBLEConfigEntry

async def async_setup_entry(
    hass: HomeAssistant,
    config_entry: HueBLEConfigEntry,
    async_add_entities: AddEntitiesCallback,
) -> None:
```

emontnemery commented on homeassistant/components/hue_ble/config_flow.py:
```diff
@@ -0,0 +1,201 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+
+STEP_USER_DATA_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_NAME): str,
+        vol.Required(CONF_MAC): str,
```
Requiring the user to manually input a MAC is not friendly, and it will also not work if our bluetooth scanners have not already found the device.

Instead, allow the user to select from a list of previously discovered matching lights, something like this:
https://github.com/home-assistant/core/blob/95b532df41a7139c8f3b24149ae3f07d3a695fdf/homeassistant/components/dormakaba_dkey/config_flow.py#L47-L92

flip-dots replied:
The primary way a user is meant to connect is through bluetooth discovery where it automatically shows up in the UI, I included this backup mechanism since I have no way of knowing if the identifier currently used to detect it is the same across all lights as I only have one model of light to test with, I can remove it if its too unfriendly.

emontnemery replied:
OK I see.

Raw input of MAC is too unfriendly IMO, it's better to update the manifest with new filters as the need arises.

emontnemery reviewed (commented):
Some initial comments.

Also, the new integration should be added in `homeassistant/brands/philips.json`

emontnemery commented:
Please click the "Ready for review" button when the comments have been addressed

flip-dots committed:
Add type

flip-dots committed:
Add type

Co-authored-by: Erik Montnemery <erik@montnemery.com>

emontnemery commented:
@flip-dots CI is not passing, please fix the failures. Be sure to follow the development guide here so you can run the linters and other checks in your development environment: https://developers.home-assistant.io/docs/development_environment

flip-dots committed:
Fix imports & run lint tool

flip-dots committed:
Run ruff format

flip-dots committed:
Regenerate requirements

flip-dots committed:
Fix lint issue in tests

emontnemery commented:
@flip-dots You still need to update `homeassistant/brands/philips.json` with the `hue_ble` integration. There's some documentation here https://developers.home-assistant.io/docs/creating_integration_brand explaining what that does

flip-dots committed:
Use specific type

flip-dots committed:
Remove support for manual configuration

flip-dots committed:
Add hue_ble as brand of phillips

flip-dots committed:
Run pylint

emontnemery reviewed (changes requested):
I think this looks pretty good now!
Please contact me on Discord if something is unclear, I'm @emontnemery there too.

flip-dots committed:
Replace state properties with state attributes

flip-dots committed:
Remove entity description

flip-dots commented on homeassistant/components/hue_ble/strings.json:
```diff
@@ -0,0 +1,20 @@
+    },
+    "error": {
+      "cannot_connect": "[%key:common::config_flow::error::cannot_connect%]",
+      "invalid_auth": "[%key:common::config_flow::error::invalid_auth%]",
```
Is it possible to put hyperlinks in error messages?. I tried putting the hyperlink to the docs that I also used for the steps but it doesn't expand.

strings.json:
```json
      "invalid_auth": "[%key:common::config_flow::error::invalid_auth%]. Ensure light is in [Alexa/Google pairing mode.]({url_pairing_mode})",
```

Result:
![image](https://github.com/user-attachments/assets/fb167b29-7f35-4f38-8f1e-a8afc0d5ab9b)

flip-dots commented on homeassistant/components/hue_ble/light.py:
```diff
@@ -0,0 +1,195 @@
+        if self._light.supports_brightness and len(supported_modes) == 0:
+            supported_modes.add(ColorMode.BRIGHTNESS)
+
+        if self._light.supports_on_off and len(supported_modes) == 0:
+            supported_modes.add(ColorMode.ONOFF)
+
+        if len(supported_modes) == 0:
+            supported_modes.add(ColorMode.UNKNOWN)
```
I would hope not, but I designed the library to be as extendible as possible. Its possible that other models use different Bluetooth attributes to be controlled and therefore might not support turning on and off using the addresses known to the library. Implementing it this way would make it easier to diagnose and hopefully later support other models (assuming that all models aren't already compatible, I only have one model to test).

flip-dots committed:
Remove should_poll

flip-dots commented on homeassistant/components/hue_ble/light.py:
```diff
@@ -0,0 +1,195 @@
+    @property
+    def should_poll(self) -> bool:
+        """Poll if light offline to trigger auto retry."""
+        return not self.available
```
Shouldn't be needed, I just added it during the early stages of development when the integration was less reliable, it works fine without it.

flip-dots committed:
Replace branch with context manager

flip-dots committed:
Add hyperlink to docs for auth error message

flip-dots committed:
Fix error when running script.hassfest

flip-dots committed:
Fix issue with light name not being shown

MartinHjelmare commented on script/hassfest/quality_scale.py:
```diff
@@ -499,6 +499,7 @@ class Rule:
     "http",
     "huawei_lte",
     "hue",
+    "hue_ble",
```
Please add a quality scale file instead. New integrations require Bronze tier:

https://developers.home-assistant.io/docs/core/integration-quality-scale/

https://developers.home-assistant.io/docs/core/integration-quality-scale/#keeping-track-of-the-implemented-rules

flip-dots replied:
I was hoping I would be able to get away with skipping that since this PR predates the introduction of quality scales but fair enough, I seem to be meeting all the requirements anyway.

flip-dots committed:
Use quality scale

bdraco commented:
Not stale. Awaits review 

Note that this likely won't get a review until the CI is passing

flip-dots committed:
Use kelvin rather than mireds

flip-dots committed:
Fix issue with external state updates not updating

flip-dots committed:
Use updated forward entry setup method

flip-dots committed:
Merge remote-tracking branch 'upstream/dev' into hue_ble

flip-dots committed:
Fix type hint

joostlek commented on homeassistant/components/hue_ble/light.py:
```diff
@@ -0,0 +1,183 @@
+        """Initialize the light object. Does not connect."""
+
+        self._light = api
+        self._name = self._light.name
```
Why do we set self._name?

flip-dots replied:
Because Hue lights have names you can set in the app and they are accessible via the API. Should I be using a different attribute for that?

joostlek commented on homeassistant/components/hue_ble/light.py:
```diff
@@ -0,0 +1,183 @@
+        self._light = api
+        self._name = self._light.name
+        self._address = self._light.address
+        self._attr_unique_id = self._light.address
+        self._attr_available = self._light.available
+        self._attr_is_on = self._light.power_state
+        self._attr_brightness = self._light.brightness
+        self._attr_color_temp_kelvin = color_util.color_temperature_mired_to_kelvin(
+            self._light.colour_temp
+        )
+        self._attr_xy_color = self._light.colour_xy
+        self._attr_min_color_temp_kelvin = color_util.color_temperature_mired_to_kelvin(
+            self._light.maximum_mireds
+        )
+        self._attr_max_color_temp_kelvin = color_util.color_temperature_mired_to_kelvin(
+            self._light.minimum_mireds
+        )
```
I would assume that this code is also used for normal updates. I would put that in a separate method to DRY

joostlek commented on homeassistant/components/hue_ble/quality_scale.yaml:
```diff
@@ -0,0 +1,60 @@
+
+  # Silver
+  action-exceptions: todo
+  config-entry-unloading: todo
+  docs-configuration-parameters: todo
+  docs-installation-parameters: todo
+  entity-unavailable: todo
+  integration-owner: todo
+  log-when-unavailable: todo
+  parallel-updates: todo
+  reauthentication-flow: todo
+  test-coverage: todo
+
+  # Gold
+  devices: todo
+  diagnostics: todo
+  discovery-update-info: todo
+  discovery: todo
+  docs-data-update: todo
+  docs-examples: todo
+  docs-known-limitations: todo
+  docs-supported-devices: todo
+  docs-supported-functions: todo
+  docs-troubleshooting: todo
+  docs-use-cases: todo
+  dynamic-devices: todo
+  entity-category: todo
+  entity-device-class: todo
+  entity-disabled-by-default: todo
+  entity-translations: todo
+  exception-translations: todo
+  icon-translations: todo
+  reconfiguration-flow: todo
+  repair-issues: todo
+  stale-devices: todo
+
+  # Platinum
+  async-dependency: todo
+  inject-websession: todo
+  strict-typing: todo
```
Please evaluate all rules

flip-dots committed:
Use await

flip-dots committed:
Use Platform.LIGHT

flip-dots committed:
Use common method for updating the updatable attributes

flip-dots committed:
Update status

joostlek commented on homeassistant/components/hue_ble/light.py:
```diff
@@ -0,0 +1,181 @@
+    async_add_entities([HaHueBLE(light)])
+
+
+class HaHueBLE(LightEntity):
```
```suggestion
class HueBLELight(LightEntity):
```

joostlek commented on homeassistant/components/hue_ble/light.py:
```diff
@@ -0,0 +1,181 @@
+        self._light = api
+        self._name = self._light.name
+        self._address = self._light.address
+        self._attr_unique_id = self._light.address
```
```suggestion
        self._attr_unique_id = api.address
```

joostlek commented on homeassistant/components/hue_ble/light.py:
```diff
@@ -0,0 +1,181 @@
+        self._attr_device_info = DeviceInfo(
+            connections={(CONNECTION_BLUETOOTH, self._light.address)},
+            manufacturer=self._light.manufacturer,
+            model=self._light.model,
```
What does this return? Is this the model or the model id?

flip-dots replied:
Model ID. For my lights its LCA006

joostlek replied:
Let's set it to the model id then

joostlek commented on homeassistant/components/hue_ble/light.py:
```diff
@@ -0,0 +1,181 @@
+
+        supported_modes = set()
+
+        if self._light.supports_colour_xy:
+            supported_modes.add(ColorMode.XY)
+
+        if self._light.supports_colour_temp:
+            supported_modes.add(ColorMode.COLOR_TEMP)
+
+        if self._light.supports_brightness and len(supported_modes) == 0:
+            supported_modes.add(ColorMode.BRIGHTNESS)
+
+        if self._light.supports_on_off and len(supported_modes) == 0:
```
Can this change on runtime? If not we can set this in the constructor

joostlek commented on homeassistant/components/hue_ble/quality_scale.yaml:
```diff
@@ -0,0 +1,60 @@
+  unique-config-entry: done
+
+  # Silver
+  action-exceptions: done
```
We currently don't catch exceptions raised by for example turning on the light

flip-dots replied:
Is it supposed to? The [docs](https://developers.home-assistant.io/docs/core/integration-quality-scale/checklist/) just say actions need to raise an exception, not catch it. If I try to turn the light on while its unplugged I get `Failed to perform the action light.turn_on. Unknown error.` in the web UI.

joostlek replied:
Sorry, rephrase, we need to rethrow the exceptions as HomeAssistantError so we can set the error message

flip-dots replied:
Fair enough, I think that's a problem for later, ill mark it as todo for now.

joostlek commented on homeassistant/components/hue_ble/strings.json:
```diff
@@ -0,0 +1,20 @@
+  "config": {
+    "step": {
+      "confirm": {
+        "description": "Do you want to set up {name} ({mac})?. Make sure the light is in [Alexa/Google pairing mode.]({url_pairing_mode})"
```
Is this pairing really called Alexa or Google pairing mode?

flip-dots replied:
The more general term they use is voice assistant pairing mode but both UIs do the same thing, they put the lights in a mode where they will accept any pair requests which is needed to control them. They will also accept pair requests when they have been factory reset but I find its easier to do it via the Hue app, that way they can still be controlled if something goes wrong and it also allows me to test that the state in HA updates when I change it in the Hue app.

| Page 1                            |   Page 2  | Page 3
| ----------------------------------- | ----------------- | ------------------ |
| ![3](https://github.com/user-attachments/assets/316a991d-2618-4a6a-a7ec-1f9babc1d56e) | ![2](https://github.com/user-attachments/assets/0c062d48-db9f-459f-910f-0d7615b2e640) | ![1](https://github.com/user-attachments/assets/2d9a32e5-e35a-409f-8df5-79cc83716c39) |

emontnemery replied:
I would suggest to make the text "Make sure the light is [made discoverable to voice assistants]." to match the UI. Then in the documentation we explain where to find this option.

joostlek commented on homeassistant/components/hue_ble/strings.json:
```diff
@@ -0,0 +1,20 @@
+  "config": {
+    "step": {
+      "confirm": {
+        "description": "Do you want to set up {name} ({mac})?. Make sure the light is in [Alexa/Google pairing mode.]({url_pairing_mode})"
```
So for example with oral B, we got a lot of comments and feedback that this caused a lot of false discoveries, for example by neighbours. Does every hue light already emit a BLE signal? Or only ones that are not connected to the hub?

flip-dots replied:
They added in Bluetooth with the 4th generation lights which came out around 2019. From my testing they always seem to broadcast via Bluetooth regardless of how they are set up.

joostlek replied:
Is there a way to decrease the amount of discoveries? As in, can we read from the light if it's connected to a bridge? Because if we tried to match every device, I would expect it to discover a lot of devices nowadays (keep in mind 2019 was 6 years ago). So the more specific we can match (as in, if we know that someone is looking to add them), the better

flip-dots replied:
Not as far as I can tell, you can get a decent amount of information even without being paired and I can see that some stuff does differ between it being connected/disconnected from a ZigBee network but I haven't been able to find a pattern that's consistent across the two different lights I have and other things like Bluetooth pairing status.

flip-dots committed:
Move supported light modes to constructor

flip-dots committed:
Update class name

flip-dots committed:
Use light name for device name

flip-dots committed:
Use different variable name for light

flip-dots committed:
Fix divide by zero when in XY colour mode

flip-dots committed:
Handle action exceptions later

flip-dots committed:
Update setup text

flip-dots committed:
Use latest HueBLE version (improved logging)

joostlek committed:
Fix

joostlek commented on tests/components/hue_ble/test_config_flow.py:
```diff
@@ -0,0 +1,240 @@
+        "url_pairing_mode": URL_PAIRING_MODE,
+    }
+
+    with patch(
+        "homeassistant.components.hue_ble.config_flow.validate_input",
+        return_value=True,
+    ):
```
We should not patch internals, instead mock out the library and make sure it responds like we expect

joostlek commented on tests/components/hue_ble/test_config_flow.py:
```diff
@@ -0,0 +1,240 @@
+        )
+        await hass.async_block_till_done()
+
+    assert result["type"] is FlowResultType.CREATE_ENTRY
+    assert result["title"] == TEST_DEVICE_NAME
+    assert result["result"].unique_id == dr.format_mac(TEST_DEVICE_MAC)
```
let's also assert `data`

joostlek commented on tests/components/hue_ble/test_config_flow.py:
```diff
@@ -0,0 +1,240 @@
+        "failed_poll",
+    ],
+)
+async def test_validation(
+    hass: HomeAssistant,
+    mock_setup_entry: AsyncMock,
+    return_device_result,
+    scanner_count_result,
+    connect_result,
+    authenticated_result,
+    connected_result,
+    poll_state_result,
+    expected_result,
+) -> None:
+    """Test input validation."""
+    with (
+        patch(
+            "homeassistant.components.hue_ble.config_flow.async_ble_device_from_address",
+            return_value=return_device_result,
+        ),
+        patch(
+            "homeassistant.components.hue_ble.config_flow.async_scanner_count",
+            return_value=scanner_count_result,
+        ),
+        patch(
+            "homeassistant.components.hue_ble.config_flow.HueBleLight",
+            autospec=True,
+            authenticated=authenticated_result,
+            connected=connected_result,
+        ) as mock_obj,
+    ):
+        client = mock_obj.return_value
+        client.connect.return_value = connect_result
+        client.authenticated = authenticated_result
+        client.connected = connected_result
+        client.poll_state.return_value = poll_state_result
+
+        with expected_result:
+            await validate_input(hass, TEST_DEVICE_MAC)
```
Yea we shouldn't test the internals like this

joostlek commented on tests/components/hue_ble/test_config_flow.py:
We should also add
1. the manual user flow
2. The manual user flow with exceptions
3. Trying to set up an already set up device
4. Same for that for Bluetooth

joostlek reviewed (changes requested):
I think it would also be nice to have light tests. I will try to get some more eyes on this PR these weeks so we can get this through

flip-dots committed:
Merge remote-tracking branch 'upstream/dev' into hue_ble

flip-dots committed:
Assert data

flip-dots committed:
Mock library in form tests

flip-dots committed:
Add test for unsupported user config flow

flip-dots committed:
Add test for already configured device

flip-dots committed:
Fix None issues

flip-dots committed:
Add tests for __init__.py

flip-dots committed:
Merge remote-tracking branch 'upstream/dev' into hue_ble

flip-dots committed:
Fix lint issue

flip-dots committed:
Add minimal test for setting of light values

flip-dots commented:
> I think it would also be nice to have light tests. I will try to get some more eyes on this PR these weeks so we can get this through

I have added some tests for turning the light on and off

flip-dots committed:
Use self._api instead of self._light

flip-dots committed:
Set model_id instead of model

flip-dots committed:
Merge remote-tracking branch 'upstream/dev' into hue_ble

emontnemery commented on homeassistant/components/hue_ble/light.py:
```diff
@@ -0,0 +1,175 @@
+"""light platform."""
```
```suggestion
"""Hue BLE light platform."""
```

emontnemery commented on homeassistant/components/hue_ble/manifest.json:
```diff
@@ -0,0 +1,19 @@
+  "iot_class": "local_push",
+  "loggers": ["bleak", "HueBLE"],
+  "quality_scale": "bronze",
+  "requirements": ["HueBLE==1.0.7"]
```
The library doesn't seem to meet all the [transparency requirements](https://developers.home-assistant.io/docs/core/integration-quality-scale/rules/dependency-transparency/), there's no PyPi release flow.

flip-dots replied:
This is a bit cheeky but can I pretty please skip this requirement 🥺, the entire quality system did not exist when I created this PR and I really don't want to have to figure out how to configure GitHub pipelines just to make it publish a library which is literally just a single Python file.

emontnemery reviewed (changes requested):
Some comments. I've not reviewed the quality scale settings or tests.

flip-dots committed:
Remove unreachable code

flip-dots committed:
Simplify validation error handling

flip-dots committed:
Log detailed errors if initial poll fails

flip-dots committed:
Remove unused class variable

flip-dots committed:
Raise error if light does not report support for turning on and off

flip-dots committed:
Update name in header docstring

flip-dots committed:
Merge remote-tracking branch 'upstream/dev' into hue_ble

flip-dots commented on homeassistant/components/hue_ble/light.py:
```diff
@@ -0,0 +1,175 @@
+        if light.supports_on_off and len(self._attr_supported_color_modes) == 0:
+            self._attr_supported_color_modes.add(ColorMode.ONOFF)
+        if len(self._attr_supported_color_modes) == 0:
+            self._attr_supported_color_modes.add(ColorMode.UNKNOWN)
```
Its a weird edge case I considered, its possible that other models use different attributes for different things (i.e what if a different model uses a different address for turning on or off but the rest is the same, and from the perspective of the underlying library it was done this way so that you would at least have some control if some of the addresses were right and to make it easier to support other Hue BLE devices (like their smart socket) or any BLE sensors they may release in future. Since I assume for Home Assistant we only want to support tested models I have now added a raise HomeAssistantError there to stop anyone from trying to use an untested device.

flip-dots committed:
Use latest HueBLE version

flip-dots committed:
Run lint tool

flip-dots committed:
Generate requirements

emontnemery commented on homeassistant/components/hue_ble/light.py:
```diff
@@ -0,0 +1,155 @@
+        if self._api.supports_brightness:
+            return ColorMode.BRIGHTNESS
+
+        return ColorMode.ONOFF
```
The logic is not correct, we will return `ColorMode.BRIGHTNESS` or `ColorMode.ONOFF` for a light which only supports `ColorMode.COLOR_TEMP`, which is not correct.

Please add a test for a light with support for `ColorMode.COLOR_TEMP` if there is not already such a test.

emontnemery reviewed (dismissed):
I think this looks mostly good now!
Some nits though. I've not looked at the tests yet.

flip-dots committed:
Move compatibility check to config flow

flip-dots committed:
Use None instead of -1

flip-dots committed:
Fix color mode logic

flip-dots committed:
Add tests for color modes

flip-dots committed:
Simplify calculation of avaliable color modes

emontnemery commented on homeassistant/components/hue_ble/config_flow.py:
```diff
@@ -0,0 +1,151 @@
+            _LOGGER.warning("Errors raised when connecting to light: %s", errors)
+            return Error.CANNOT_CONNECT
+        else:  # noqa: RET505
+            return None
```
Why do you ignore the lint error?
I'd suggest to adjust the code instead:
```suggestion
        return None
```

flip-dots committed:
Use try else

flip-dots committed:
Simplify color mode validation

joostlek committed:
Merge branch 'dev' into hue_ble

joostlek committed:
fix
